crates/nu-cli/src/types/deduction.rs

```diff
@@ -0,0 +1,1451 @@
+        bin_spanned: &SpannedExpression,
+        (pipeline_idx, pipeline): (usize, &Commands),
+        registry: &CommandRegistry,
+    ) -> Result<(), ShellError> {
+        trace!("Infering shapes between var {:?} and expr {:?}", var, expr);
+        let bin = match &bin_spanned.expr {
+            Expression::Binary(bin) => bin,
+            _ => unreachable!(),
+        };
+        if let Expression::Literal(Literal::Operator(op)) = bin.op.expr {
+            match &op {
+                //For || and && we insert shapes decay able to bool
+                Operator::And | Operator::Or => {
+                    let shapes = get_shapes_decay_able_to_bool();
+                    // shapes.push(SyntaxShape::Math);
+                    self.checked_insert(
+                        &var,
+                        VarShapeDeduction::from_usage_with_alternatives(&var.span, &shapes),
+                    )?;
+                }
+                Operator::Contains | Operator::NotContains => {
+                    self.checked_insert(
+                        var,
+                        vec![VarShapeDeduction::from_usage(
+                            &var.span,
+                            &SyntaxShape::String,
+                        )],
+                    )?;
+                }
+                Operator::In | Operator::NotIn => {
+                    match var_side {
+                        BinarySide::Left => match &expr.expr {
+                            Expression::List(list) => {
+                                if !list.is_empty() {
+                                    let shapes_in_list = self
+                                        .get_shapes_in_list_or_insert_dependency(
+                                            var,
+                                            bin_spanned,
+                                            &list,
+                                            (pipeline_idx, pipeline),
+                                            registry,
+                                        );
+                                    match shapes_in_list {
+                                        None => {}
+                                        Some(shapes_in_list) => {
+                                            self.checked_insert(
+                                                var,
+                                                VarShapeDeduction::from_usage_with_alternatives(
+                                                    &var.span,
+                                                    &shapes_in_list,
+                                                ),
+                                            )?;
+                                        }
+                                    }
+                                }
+                            }
+                            //REVIEW is var in table legal? Shouldn't the rhs in "$p in [a b c]" be 
+                            //parsed as a list?
```

Comment:
We've got a few things to clean up (not as part of this PR) in the names of things. The Table below should actually be an inner table. You're right, that in your example that should be a list.